### PR TITLE
Allow `SessionContext::read_csv`, etc to read multiple files

### LINKED ISSUE #4909: Read multiple files/folders using `read_csv`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
I want to load files from multiple locations. In Spark, I can just give a list of paths and it works. Would like to do the same with read_csv, read_avro and read_parquet methods.

**Describe the solution you'd like**
Read a list of files using methods like so: read_csv(vec![String::new()])

**Describe alternatives you've considered**
I can read the paths one by one and then concatenate the dataframes but not sure if it is ok to do so.

**Additional context**
I have files from S3 in multiple locations and use the above methods to read them.

### PR DESCRIPTION
**Pull request by saikrishna1-bidgely**
closes #4909

### CODE REVIEW

**Comment by saikrishna1-bidgely** [comment]
I'll add for `read_avro`, `read_json` and `read_parquet` too once what needs to be done gets finalised.

File: datafusion/core/src/execution/context.rs
Comment (alamb) [inline datafusion/core/src/execution/context.rs:554]: what do you think about calling this `read_avro_from_paths`? rather than `multi_paths`?

File: datafusion/core/src/execution/context.rs
Comment (alamb) [inline datafusion/core/src/execution/context.rs:597]: Perhaps we can update the docstring as well?

**Review by alamb — COMMENTED** [review]
This looks like a nice improvement @saikrishna1-bidgely 

I think we should add a test for this new functionality so that we don't accidentally break the new APIs going forward. 

Also I was wondering about the signature

Rather than a slice, what would you think about taking something that could be turned into an iter:

So instead of 

```rust
    pub async fn read_parquet_with_multi_paths(
        &self,
        table_paths: &[impl AsRef<str>],
        options: ParquetReadOptions<'_>,
    ) -> Result<DataFrame> {
```

Something more like

```rust
    pub async fn read_parquet_with_multi_paths(
        &self,
        table_paths: impl IntoIterator<Item = &str>],
        options: ParquetReadOptions<'_>,
    ) -> Result<DataFrame> {
```

I also think it would be ideal to figure out some way to have the same API take both a single path and an iterator -- do you think the above would work?

**Comment by alamb** [comment]
> What do you think about having a single method which only takes a list of paths? For a single path, the callee can create a slice/Vec. This would be a lot simpler to do.

I was thinking about this PR and I have an alternate suggestion

It seems to me that  `read_parquet`, `read_avro`, etc are wrappers to simplify the process of creating a `ListingTable`. Support for multiple paths starts complicating the API more -- what do you think about instead of adding `read_parquet_from_path`s we make it easier to see how to read multiple files using the `ListingTable` API directly?

For example, I bet if we added a doc example like the following

```rust
    /// Creates a [`DataFrame`] for reading a Parquet data source from a single file or directory. 
    ///
    /// Note: if you want to read from multiple files, or control other behaviors
    /// you can use the [`ListingTable`] API directly. For example to read multiple files
    /// 
    /// ```
    /// Example here (basically copy/paste the implementation of read_parquet and support multiple files)
    /// ```
    pub async fn read_parquet(
        &self,
        table_path: impl AsRef<str>,
        options: ParquetReadOptions<'_>,
    ) -> Result<DataFrame> {
...
```

We could give similar treatment to the docstrings for `read_avro` and `read_csv` (perhaps by pointing to the docs for `read_parquet` for an example of creating `ListingTable`s)

**Comment by saikrishna1-bidgely** [comment]
Quick question, wouldn't we want to support multiple paths anyways since we would want to use it in DataFusion-CLI?

Also, if we are able to crack the implementation which can take both single and multiple paths in the same function, API itself should be unchanged, right?

So, I think the following are our options:
1. As you said, don't support it but provide an example in the docs.
2. Have multiple methods.
3. Same method takes both single and multiple paths.
4. Change the current methods to take only multiple paths.

1 is definitely the simplest but I think we should support it as it would make using DataFusion simpler.
2 is simple but multiple functions is a downside.
3 if possible is best case IMO.
4 is also simpler but needs a lot of changes downstream.

**Comment by alamb** [comment]
> Quick question, wouldn't we want to support multiple paths anyways since we would want to use it in DataFusion-CLI?

I think that is likely a separate question -- DataFusion CLI can potentially create a ListingTable directly rather than using the higher level SessionContext API as well

> Also, if we are able to crack the implementation which can take both single and multiple paths in the same function, API itself should be unchanged, right?

Yes, I agree 

> So, I think the following are our options:

I agree with your assessments -- I do think documentation on how to create a ListingTable would go a long way. It seems we are lacking in such docs now https://docs.rs/datafusion/16.0.0/datafusion/datasource/listing/struct.ListingTable.html

I think documentation will help regardless of what else we chose to do -- I'll go write some now. Thank you for this good discussion

**Comment by saikrishna1-bidgely** [comment]
@alamb I finally got a way to implement overloading using Traits.
```rust
use std::vec;

fn main() {
    struct PATH {
    }
    
    impl PATH {
        pub fn new() -> Self {
            Self {
            }
        }
    }
    
    pub trait Reader<T>: Sized {
        fn read_csv(&self, value: T) -> i32;
    }
    
    impl<'a> Reader<&'a str> for PATH {
        fn read_csv(&self, p: &'a str) -> i32  {
            self.read_csv(vec![p])
        }
    }
    
    impl<'a> Reader<Vec<&'a str>> for PATH {
        fn read_csv(&self, v: Vec<&'a str>) -> i32 {
            v.len() as i32
        }
    }
    let p = PATH::new();
    println!("{:?}", p.read_csv("path"));
    println!("{:?}", p.read_csv(vec!["path", "paths2"]));
}
```
This way, callees using `str` or `String` can continue using the function and we can add support for vector/iterators. I will write a more general solution for read functions.

I tried to implement using Enum and Union but I wasn't able to do so and they will change the function signature.

**Comment by alamb** [comment]
Thank you @saikrishna1-bidgely  -- sounds like great progress. I think if we go the trait route as long as we document how to use it (basically we can make a doc example showing the use of a &str) I think we'll be good. 

Thanks again!

File: datafusion/core/src/execution/context.rs
Comment (saikrishna1-bidgely) [inline datafusion/core/src/execution/context.rs:1055]: We can't use AsRef<str> here. We have to implement both `&str` and `String` seperately.

File: datafusion/core/src/execution/context.rs
Comment (saikrishna1-bidgely) [inline datafusion/core/src/execution/context.rs:1057]: For now implemented only for `read_csv` but once it gets finalised for it, I'll implement for the rest of the methods.

File: datafusion/core/src/execution/context.rs
Comment (saikrishna1-bidgely) [inline datafusion/core/src/execution/context.rs:1078]: Should we implement a trait `ReadOptions`?
```rust
trait ReadOptions {
    fn to_listing_options(&self, target_partitions: usize) -> ListingOptions;
    fn get_resolved_schema(&self) -> Result<SchemaRef>;
}
```
`CsvReadOptions` and other will implement this trait. This way, we can avoid repeating this code for other methods.

Comment (saikrishna1-bidgely) [reply datafusion/core/src/execution/context.rs:1085]: Now implemented in #5025.

**Comment by alamb** [comment]
Marking as a draft so it is clear this is waiting on some updates rather than waiting on review

**Comment by alamb** [comment]
Please mark it ready for review once if I made a mistake or when it is next ready

File: datafusion/core/src/dataframe.rs
Comment (alamb) [inline datafusion/core/src/dataframe.rs:65]: I am somewhat worried about session context getting a new trait like this as now everyone who uses SessionContext must use now add this new `use` statement

File: datafusion/core/src/execution/context.rs
Comment (alamb) [inline datafusion/core/src/execution/context.rs:646]: I wonder if you could leave these methods on `SessionContext` and instead use a trait for the argument and avoid changes to downstream users. 

Perhaps something like:


```rust
trait DataFilePaths {
  // Parse to a list of URs
  fn to_urls(self) -> Result<Vec<ListingTableUrl>>>;
}

impl DataFilePaths for &str {
  fn to_urls(self) -> Result<Vec<ListingTableUrl>>> {
    Ok(vec![ListingTableUrl::parse(self)?])
  }
}

impl DataFilePaths for Vec<&str> {
  fn to_urls(self) -> Result<Vec<ListingTableUrl>>> {
    self
            .iter()
            .map(ListingTableUrl::parse)
            .collect::<Result<Vec<ListingTableUrl>>>()
  }
}



impl SessionContext {
...
    pub async fn <P> read_avro(
        &self,
        table_path: P,
        options: AvroReadOptions<'_>,
    ) -> Result<DataFrame> {
        self._read_type(table_path.to_urls()>, options).await
    }
...
```

**Review by alamb — COMMENTED** [review]
Thank you @saikrishna1-bidgely  -- I am a little concerned about the changes required for users of datafusion if we merged this PR. I had a suggestion that might make migration easier. 

Any thoughts @andygrove  or others?

File: datafusion/core/src/execution/context.rs
Comment (saikrishna1-bidgely) [reply datafusion/core/src/execution/context.rs:646]: This might be better. Less code to implement. But we have to use the `use` statement for the defined trait whenever we call `read_csv` and others.

One more issue could be when the users implement a trait with the same method name. They won't be able to use both traits, only one.

Comment (saikrishna1-bidgely) [reply datafusion/core/src/execution/context.rs:646]: Hi @alamb, I was travelling and couldn't work on this last week. In the above example, we will need to use the `use` statement even when we are using `read_csv("s3://bucket/key", options)`. Instead I propose the following:
1. What if we implement AsRef for Vec. Then, we only need to use the `use` statement when we are calling for multiple  paths. The default single path code will remain unchanged. The AsRef will simply concatenate all the strings in Vec with `,` delimiter. And in the `_read_type`, we will split on `,`.
2. How about having a different method `read_csvs` for taking multiple paths. `read_csvs` is a lot cleaner than `read_csv_with_multi_paths`.

**Comment by saikrishna1-bidgely** [comment]
Hi @alamb I was on vacation and couldn't work on this. I've updated the comments pls go through them once.

**Comment by alamb** [comment]
Thanks @saikrishna1-bidgely  -- will check them out

File: datafusion/core/src/execution/context.rs
Comment (alamb) [reply datafusion/core/src/execution/context.rs:646]: > How about having a different method read_csvs for taking multiple paths. read_csvs is a lot cleaner than read_csv_with_multi_paths.

I agree that would be better

I think my concern was about the fact that then there would be `read_parquets` `read_avros`, etc. which starts to sound like too many 🤔

Comment (saikrishna1-bidgely) [reply datafusion/core/src/execution/context.rs:646]: Looks like a toss up between multiple methods and using traits. Shall we get inputs from other core members?

Comment (alamb) [reply datafusion/core/src/execution/context.rs:646]: Yes, I think that would be a good idea. Perhaps @andygrove @Dandandan @yahoNanJing @liukun4515  or @tustvold  have some thoughts

Comment (tustvold) [reply datafusion/core/src/execution/context.rs:646]: > But we have to use the use statement for the DataFilePaths whenever we call read_csv and others.

I'm surprised by this, you shouldn't need to use a trait in order to satisfy a type constraint on a generic method?

Comment (saikrishna1-bidgely) [reply datafusion/core/src/execution/context.rs:646]: The generic method would have a constraint on P as `P: DataFilePaths`. When we call the method `read_csv("path", options)` we will have to use the `use ..DataFilePaths` so that rust knows the trait `DataFilePaths` is implemented for string. Is this what you meant?

Comment (tustvold) [reply datafusion/core/src/execution/context.rs:646]: That isn't how traits behave, you do not need to have a trait in scope in order to call a generic method with a type constraint on it - https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a629bf1bdd4b6418ab09e274f8214de5

Comment (saikrishna1-bidgely) [reply datafusion/core/src/execution/context.rs:646]: Thanks. I didn't know traits worked that way with generics. I tried an implementation and it worked!

**Comment by saikrishna1-bidgely** [comment]
@alamb @tustvold I updated the code with suggestion from @tustvold. The change is now much smaller and cleaner. Pls review the code.

File: datafusion/core/src/execution/context.rs
Comment (alamb) [inline datafusion/core/src/execution/context.rs:672]: ❤️

File: datafusion/core/src/execution/context.rs
Comment (alamb) [inline datafusion/core/src/execution/context.rs:2800]: Are these methods still needed?

File: datafusion/core/src/execution/context.rs
Comment (alamb) [inline datafusion/core/src/execution/context.rs:111]:
```suggestion
/// DataFilePaths adds a method to convert strings and vector of strings to vector of [`ListingTableUrl`] URLs.
/// This allows methods such [`SessionContext::read_csv`] and `[`SessionContext::read_avro`]
/// to take either a single file or multiple files. 
```

File: datafusion/core/src/execution/context.rs
Comment (alamb) [inline datafusion/core/src/execution/context.rs:117]: technically speaking I think this might be a breaking change as the methods used to take `impl AsRef<str>` 

However, I think it is fine for this case as there is an easy workaround (call `as_ref()`)

**Review by alamb — COMMENTED** [review]
Thank you @saikrishna1-bidgely  -- I really like this PR ❤️ -- thank you for sticking with it.

The only thing I think is missing is a test (mostly to ensure this ability is not removed by accident in the future). 

For a test, What do you think about
1.  adding a doc example on one of the methods (e.g like `SessionContext::read_parquet` that shows reading from a list of strings)  
2. Add a note on the other methods (e.g. `SessionContext::read_csv`) pointing at the method with the example.

I can help with this documentation if you like

File: datafusion/core/src/execution/context.rs
Comment (saikrishna1-bidgely) [reply datafusion/core/src/execution/context.rs:117]: I'm trying to implement for AsRef<str> along with an iterator. I'll update once it's ready.

**Comment by saikrishna1-bidgely** [comment]
Regarding the docs, I think we should extend the example in `SessionContext` and add an example to all the `read_*` methods.

File: datafusion/core/src/execution/context.rs
Comment (saikrishna1-bidgely) [reply datafusion/core/src/execution/context.rs:117]: Looks like rust won't allow it. We won't be able to implement a trait for another trait and also for another class.

**Comment by saikrishna1-bidgely** [comment]
@alamb updated the docs. Should I remove `call_read_csvs` and the related methods?

**Comment by saikrishna1-bidgely** [comment]
@alamb removed methods from `CallReadTrait` too.

File: datafusion/core/src/execution/context.rs
Comment (alamb) [inline datafusion/core/src/execution/context.rs:164]: I think this particular line is just confusing -- the next line ignores the result (redefines `df`) and it registers the same file twice. Given this is supposed to be the simplest example, I think we should leave this one alone.

File: datafusion/core/src/execution/context.rs
Comment (alamb) [inline datafusion/core/src/execution/context.rs:736]:
```suggestion
    /// // You can read a single file using `read_csv`
    /// let df = ctx.read_csv("tests/data/example.csv", CsvReadOptions::new()).await?;
    /// // you can also read multiple files:
    /// let df = ctx.read_csv(vec!["tests/data/example.csv", "tests/data/example.csv"], CsvReadOptions::new()).await?;
```

**Review by alamb — COMMENTED** [review]
Thanks @saikrishna1-bidgely  I think just a few more comment updates and this will be ready to go. Thanks for sticking with it

File: datafusion/core/src/execution/context.rs
Comment (saikrishna1-bidgely) [reply datafusion/core/src/execution/context.rs:164]: Makes sense. So, let's just extend `read_csv` example?

**Review by alamb — APPROVED** [review]
Thank you @saikrishna1-bidgely  -- I think this is looking ready to go.

**Comment by saikrishna1-bidgely** [comment]
Cool! Do we need more approvals before merging?

**Comment by alamb** [comment]
> Cool! Do we need more approvals before merging?

Nope, I was just giving it some time after approval for other maintainers to have a look if they wanted. 

🚀

**Comment by alamb** [comment]
Closing/ reopening to rerun CI (for some reason several of the tests were canceled)

**Comment by alamb** [comment]
Thanks again for sticking with this @saikrishna1-bidgely

**Comment by alamb** [comment]
Thanks again for sticking with this @saikrishna1-bidgely

**Pull request merged**
